File: TestUCI.c
#include <windows.h>
#include <stdlib.h>
#include <stdio.h>

typedef int (__stdcall* F_UCIDecode)(const void *src, int srclen, void** dst, int* stride, int* w, int* h, int* b);
typedef int (__stdcall* F_UCIFree)(void* p);
typedef int (__stdcall* F_UCIDebug)(int level);

int main()
{
	unsigned char *src, *dst;
	int w, h, b, r, srclen, stride, i;
	FILE* fp;

	HMODULE hucimod = LoadLibrary("ucidec.dll");
	F_UCIDecode UCIDecode;
	F_UCIFree UCIFree;
	F_UCIDebug UCIDebug;
	if(!hucimod)
	{
		printf("ERROR: can not load ucidec.dll\n");
		return -1;
	}

	UCIDecode = (F_UCIDecode)GetProcAddress(hucimod, "UCIDecode");
	UCIFree = (F_UCIFree)GetProcAddress(hucimod, "UCIFree");
	UCIDebug = (F_UCIDebug)GetProcAddress(hucimod, "UCIDebug");
	if(!UCIDecode || !UCIFree || !UCIDebug)
	{
		printf("ERROR: can not load UCIDecode or UCIFree or UCIDebug from ucidec.dll\n");
		return -2;
	}

	fp = fopen("test.uci", "rb");
	if(!fp)
	{
		printf("ERROR: can not open test.uci\n");
		return -3;
	}

	fseek(fp, 0, SEEK_END);
	srclen = ftell(fp);
	fseek(fp, 0, SEEK_SET);
	src = (unsigned char*)malloc(srclen);
	fread(src, 1, srclen, fp);
	fclose(fp);

	UCIDebug(0x7fffffff);
	r = UCIDecode(src, srclen, (void**)&dst, &stride, &w, &h, &b);
	free(src);
	if(r < 0)
	{
		printf("ERROR: UCIDecode failed (return %d)\n", r);
		return -4;
	}
	printf("INFO: width x height x bit : %d x %d x %d\n", w, h, b);

	fp = fopen("test.rgb", "wb");
	if(!fp)
	{
		UCIFree(dst);
		printf("ERROR: can not create test.rgb\n");
		return -5;
	}
	for(i = 0; i < h; ++i)
		fwrite(dst + i * stride, 1, w * (b/8), fp);
	fclose(fp);

	UCIFree(dst);

	return 0;
}

File: TestUCI.lua
-- for LuaJIT 2.0-beta6 or later

local ffi = require("ffi")
ffi.cdef[[
	int  __stdcall UCIDecode(const void* src, int srclen, void** dst, int* stride, int* width, int* height, int* bit);
	void __stdcall UCIFree(void* p);
	void __stdcall UCIDebug(int level);
]]
local ucidec = ffi.load("ucidec")

if not ucidec then
	print "ERROR: can not load ucidec.dll"
	return -1
end

local fsrc = io.open("test.uci", "rb")
if not fsrc then
	print "ERROR: can not open test.uci"
	return -2
end

local src = fsrc:read("*a")
fsrc:close()
if not src then
	print "ERROR: can not read test.uci"
	return -3
end

local dst = ffi.new("void*[1]")
local stride = ffi.new("int[1]")
local w = ffi.new("int[1]")
local h = ffi.new("int[1]")
local b = ffi.new("int[1]")

ucidec.UCIDebug(0x7fffffff)
local r = ucidec.UCIDecode(src, #src, dst, stride, w, h, b)
if r < 0 then
	print(string.format("ERROR: DecodeUCI failed (return %d)", r))
	return -4
end
print(string.format("INFO: width x height x bit : %d x %d x %d", w[0], h[0], b[0]))

local fdst = io.open("test.rgb", "wb")
if not fdst then
	print "ERROR: can not create test.rgb"
	return -5
end

local dstbase = ffi.cast("char*", dst[0])
local linesize = w[0] * (b[0] / 8)
for i = 1, h[0] do
	fdst:write(ffi.string(dstbase + i * stride[0], linesize))
end
fdst:close()

ucidec.UCIFree(dst[0])

return 0
